diff -urN original/e2fsprogs-1.41.12/config/config.sub build/e2fsprogs-1.41.12/config/config.sub
--- original/e2fsprogs-1.41.12/config/config.sub
+++ build/e2fsprogs-1.41.12/config/config.sub
@@ -288,6 +288,7 @@
 	| sparcv8 | sparcv9 | sparcv9b | sparcv9v \
 	| spu | strongarm \
 	| tahoe | thumb | tic4x | tic80 | tron \
+	| ubicom32 \
 	| v850 | v850e \
 	| we32k \
 	| x86 | xc16x | xscale | xscalee[bl] | xstormy16 | xtensa \
@@ -374,6 +375,7 @@
 	| tahoe-* | thumb-* \
 	| tic30-* | tic4x-* | tic54x-* | tic55x-* | tic6x-* | tic80-* | tile-* \
 	| tron-* \
+	| ubicom32-* \
 	| v850-* | v850e-* | vax-* \
 	| we32k-* \
 	| x86-* | x86_64-* | xc16x-* | xps100-* | xscale-* | xscalee[bl]-* \
diff -urN original/e2fsprogs-1.41.12/e2fsck/Makefile.in build/e2fsprogs-1.41.12/e2fsck/Makefile.in
--- original/e2fsprogs-1.41.12/e2fsck/Makefile.in
+++ build/e2fsprogs-1.41.12/e2fsck/Makefile.in
@@ -196,7 +196,7 @@
 	$(Q) $(MKINSTALLDIRS) $(DESTDIR)$(root_sbindir) \
 		$(DESTDIR)$(man8dir) $(DESTDIR)$(man5dir)
 
-install: $(PROGS) $(MANPAGES) $(FMANPAGES) installdirs
+install: $(PROGS)
 	$(Q) for i in $(PROGS); do \
 		echo "	INSTALL $(root_sbindir)/$$i"; \
 		$(INSTALL_PROGRAM) $$i $(DESTDIR)$(root_sbindir)/$$i; \
@@ -206,25 +206,6 @@
 		(cd $(DESTDIR)$(root_sbindir); \
 			$(LN) $(LINK_INSTALL_FLAGS) e2fsck fsck.$$i); \
 	done
-	$(Q) for i in $(MANPAGES); do \
-		for j in $(COMPRESS_EXT); do \
-			$(RM) -f $(DESTDIR)$(man8dir)/$$i.$$j; \
-		done; \
-		echo "	INSTALL_DATA $(man8dir)/$$i"; \
-		$(INSTALL_DATA) $$i $(DESTDIR)$(man8dir)/$$i; \
-	done
-	$(Q) for i in $(FMANPAGES); do \
-		for j in $(COMPRESS_EXT); do \
-			$(RM) -f $(DESTDIR)$(man5dir)/$$i.$$j; \
-		done; \
-		echo "	INSTALL_DATA $(man5dir)/$$i"; \
-		$(INSTALL_DATA) $$i $(DESTDIR)$(man5dir)/$$i; \
-	done
-	$(Q) for i in ext2 ext3 ext4 ext4dev; do \
-		echo "	LINK $(man8dir)/fsck.$$i.8"; \
-		(cd $(DESTDIR)$(man8dir); \
-			$(LN) $(LINK_INSTALL_FLAGS) e2fsck.8 fsck.$$i.8); \
-	done
 
 install-strip: install
 	$(Q) for i in $(PROGS); do \
diff -urN original/e2fsprogs-1.41.12/misc/Makefile.in build/e2fsprogs-1.41.12/misc/Makefile.in
--- original/e2fsprogs-1.41.12/misc/Makefile.in
+++ build/e2fsprogs-1.41.12/misc/Makefile.in
@@ -23,9 +23,8 @@
 @BLKID_CMT@FINDFS_LINK= findfs
 @BLKID_CMT@FINDFS_MAN= findfs.8
 
-SPROGS=		mke2fs badblocks tune2fs dumpe2fs $(BLKID_PROG) logsave \
-			$(E2IMAGE_PROG) @FSCK_PROG@ e2undo
-USPROGS=	mklost+found filefrag e2freefrag $(UUIDD_PROG)
+SPROGS=		mke2fs
+USPROGS=	
 SMANPAGES=	tune2fs.8 mklost+found.8 mke2fs.8 dumpe2fs.8 badblocks.8 \
 			e2label.8 $(FINDFS_MAN) $(BLKID_MAN) $(E2IMAGE_MAN) \
 			logsave.8 filefrag.8 e2freefrag.8 e2undo.8 $(UUIDD_MAN) @FSCK_MAN@
@@ -394,7 +393,7 @@
 		$(DESTDIR)$(man1dir) $(DESTDIR)$(man5dir) \
 		$(DESTDIR)$(libdir) $(DESTDIR)/$(root_sysconfdir)
 
-install: all $(SMANPAGES) $(UMANPAGES) installdirs
+install: all
 	$(Q) for i in $(SPROGS); do \
 		echo "	INSTALL $(root_sbindir)/$$i"; \
 		$(INSTALL_PROGRAM) $$i $(DESTDIR)$(root_sbindir)/$$i; \
@@ -408,49 +407,6 @@
 		(cd $(DESTDIR)$(root_sbindir); \
 			$(LN) $(LINK_INSTALL_FLAGS) mke2fs mkfs.$$i); \
 	done
-	$(Q) (cd $(DESTDIR)$(root_sbindir); \
-		$(LN) $(LINK_INSTALL_FLAGS) tune2fs e2label)
-	$(Q) if test -n "$(FINDFS_LINK)"; then \
-		echo "	LINK $(root_sbindir)/findfs"; \
-		(cd $(DESTDIR)$(root_sbindir); \
-		$(LN) $(LINK_INSTALL_FLAGS) tune2fs $(FINDFS_LINK)); \
-	fi
-	$(Q) for i in $(UPROGS); do \
-		echo "	INSTALL $(bindir)/$$i"; \
-		$(INSTALL_PROGRAM) $$i $(DESTDIR)$(bindir)/$$i; \
-	done
-	$(Q) for i in $(LPROGS); do \
-		echo "	INSTALL $(libdir)/$$i"; \
-		$(INSTALL_PROGRAM) $$i $(DESTDIR)$(libdir)/$$i; \
-	done
-	$(Q) for i in $(SMANPAGES); do \
-		for j in $(COMPRESS_EXT); do \
-			$(RM) -f $(DESTDIR)$(man8dir)/$$i.$$j; \
-		done; \
-		echo "	INSTALL_DATA $(man8dir)/$$i"; \
-		$(INSTALL_DATA) $$i $(DESTDIR)$(man8dir)/$$i; \
-	done
-	$(Q) $(RM) -f $(DESTDIR)$(man8dir)/mkfs.ext2.8.gz \
-		$(DESTDIR)$(man8dir)/mkfs.ext3.8.gz
-	$(Q) for i in ext2 ext3 ext4 ext4dev; do \
-		echo "	LINK mkfs.$$i.8"; \
-		(cd $(DESTDIR)$(man8dir); \
-			$(LN) $(LINK_INSTALL_FLAGS) mke2fs.8 mkfs.$$i.8); \
-	done
-	$(Q) for i in $(UMANPAGES); do \
-		for j in $(COMPRESS_EXT); do \
-			$(RM) -f $(DESTDIR)$(man1dir)/$$i.$$j; \
-		done; \
-		echo "	INSTALL_DATA $(man1dir)/$$i"; \
-		$(INSTALL_DATA) $$i $(DESTDIR)$(man1dir)/$$i; \
-	done
-	$(Q) for i in $(FMANPAGES); do \
-		for j in $(COMPRESS_EXT); do \
-			$(RM) -f $(DESTDIR)$(man5dir)/$$i.$$j; \
-		done; \
-		echo "	INSTALL_DATA $(man5dir)/$$i"; \
-		$(INSTALL_DATA) $$i $(DESTDIR)$(man5dir)/$$i; \
-	done
 	$(Q) if test -f $(DESTDIR)$(root_sysconfdir)/mke2fs.conf; then \
 		if cmp -s $(DESTDIR)$(root_sysconfdir)/mke2fs.conf \
 			$(srcdir)/mke2fs.conf; then \
