diff -urN /tmp/build/ori/sqlite-autoconf-3070400/config.sub /tmp/build/sqlite-autoconf-3070400/config.sub
--- ori/sqlite-autoconf-3070400/config.sub	2008-11-20 02:13:36.000000000 +0800
+++ sqlite-autoconf-3070400/config.sub	2011-05-25 22:08:09.422283530 +0800
@@ -270,6 +270,7 @@
 	| sparcv8 | sparcv9 | sparcv9b \
 	| strongarm \
 	| tahoe | thumb | tic4x | tic80 | tron \
+	| ubicom32 \
 	| v850 | v850e \
 	| we32k \
 	| x86 | xscale | xscalee[bl] | xstormy16 | xtensa \
@@ -352,6 +353,7 @@
 	| tahoe-* | thumb-* \
 	| tic30-* | tic4x-* | tic54x-* | tic55x-* | tic6x-* | tic80-* \
 	| tron-* \
+	| ubicom32-* \
 	| v850-* | v850e-* | vax-* \
 	| we32k-* \
 	| x86-* | x86_64-* | xps100-* | xscale-* | xscalee[bl]-* \

